elastic/kibana
# [Embeddable Rebuild] Make React embeddables work in Canvas

Closes https://github.com/elastic/kibana/issues/179548

## Summary

This PR makes it so that React embeddables will now work in Canvas. It does so by doing two main things:

1. It ensures that the `ReactEmbeddableRenderer` is used in Canvas when an embeddable exists in the React embeddable registry - if it does not exist, it continues to use the legacy embeddable factory's `render` method.

   Since Canvas auto-applies all changes and doesn't have save functionality like Dashboard does, we must keep track of changes **as they happen** and update the expression to match the new Embeddable input - therefore, I had to add a `onAnyStateChange` callback to the `ReactEmbeddableRenderer` as a backdoor for Canvas. As a bonus to this, embeddables that **previously** didn't respect inline editing (such as the Image embeddable) will start to work once they are converted! 


3. It adds a new trigger (`ADD_CANVAS_ELEMENT_TRIGGER`) specifically for registering an embeddable to the **Canvas** add panel menu. This trigger can be attached to the **same action** that the Dashboard `ADD_PANEL_TRIGGER` is attached to - this makes it super simple to add React embeddables to Canvas:
    
    ```typescript
    uiActions.registerAction<EmbeddableApiContext>({
      id: ADD_EMBEDDABLE_ACTION_ID,
      isCompatible: async ({ embeddable }) => { ... },
      execute: async ({ embeddable }) => { ... },
      getDisplayName: () => { ... },
    });

    // register this action to the Dashboard add panel menu:
    uiActions.attachAction('ADD_PANEL_TRIGGER', ADD_EMBEDDABLE_ACTION_ID);  

    // register this action to the Canvas add panel menu:
    uiActions.attachAction('ADD_CANVAS_ELEMENT_TRIGGER', ADD_EMBEDDABLE_ACTION_ID);
    ```

As a small cleanup, I also replaced some inline embeddable expressions with `embeddableInputToExpression` - this is because I was originally missing `| render` at the end of my expression, and I didn't catch it because the expressions I was comparing it to were all declared inline. This should help keep things more consistent.

### How to Test
I attached the `ADD_EUI_MARKDOWN_ACTION_ID` action to the new `ADD_CANVAS_ELEMENT_TRIGGER`, so the new EUI Markdown React embeddable should show up in the Canvas add panel menu. Ensure that this React embeddable can be added to a workpad, and make sure it responds to edits as expected:


https://github.com/elastic/kibana/assets/8698078/d9062949-9c61-4c9e-8a8e-08042753eab6



### Checklist


### For maintainers
